app/services/query-service.js

```diff
@@ -6,13 +6,85 @@ const { createHistoryAwareRetriever } = require('langchain/chains/history_aware_
 const { FakeChatModel } = require('@langchain/core/utils/testing')
 const { AzureAISearchVectorStore } = require('../lib/azure-vector-store')
 const config = require('../config')
+const { logEvent } = require('../insights')
 
 const onFailedAttempt = async (error) => {
   if (error.retriesLeft === 0) {
     throw new Error(`Failed to get embeddings: ${error}`)
   }
 }
 
+const validateResponseLinks = (response) => {
+  let errorMessage = 'validateResponseLinks failed because no response object provided'
+  if (!response) {
+    logEvent(errorMessage)
+
+    return errorMessage
+  }
+
+  const extractLinks = (jsonObj) => {
+    const entries = []
+
+    try {
+      const traverse = (obj, parent) => {
```

Comment:
This is interesting but I don't think we need to search through the entire json tree, given we know what the shape of the json object will look like. This could get very slow if we choose to expand the tree later on. You could simply concatenate all the pageContent strings together and call `.includes()` with the generated url. 

and to make it even clearer, instead of taking the entire pageContent, you could just pluck everything before the `===` identifier

response.answer
```JSON
{
	"answer": "...",
	"items": [
		{
			"title": "...",
			"scheme": "...",
			"url": "https://www.gov.uk/countryside-stewardship-grants/temporary-deer-fencing-fg10",
			"summary": "..."
		}
	]
}
```

response.context:
```JSON
[
	{
		"pageContent": "(Title: FG10: Temporary deer fencing | Grant Scheme Name: Countryside Stewardship (CS) | Source: https://www.gov.uk/countryside-stewardship-grants/temporary-deer-fencing-fg10 | Chunk Number: 0)===## How much will be paid  £8.09 per metre."
	}
]
```